### PR TITLE
[ROCm][DeepSeek-V4] WIP: Enable CSA multistream decode

### PR DESCRIPTION
**Pull request by Fangzhou-Ai**
Addresses #41820.

## Summary

This PR enables ROCm DeepSeek-V4 CSA multistream decode.

Changes:
- Adds ROCm CSA multistream scheduling for DeepSeek-V4 decode.
- Splits q/KV post-RMSNorm work so KV cache insert, compressor, and indexer work can run on auxiliary streams.
- Uses ROCm defaults: `strategy=overlap`, graph modes `none,piecewise`, split q/KV post path enabled, deferred projections disabled.
- Applies ROCm CSA multistream branch scheduling only when the multistream scheduler is active for the current step.
- Adds defensive bounds masking in ROCm AITER sparse MLA helpers.

## Duplicate Work Check

I checked:
- `gh issue view 41820 --repo vllm-project/vllm --comments`
- `gh pr list --repo vllm-project/vllm --state open --search "41820 in:body"`
- `gh pr list --repo vllm-project/vllm --state open --search "DeepSeek V4 ROCm"`
- `gh pr list --repo vllm-project/vllm --state open --search "DSV4 CSA ROCm"`

Related open ROCm/DSV4 PRs exist, including #41136, #41601, #42908, #43306, and #43679. I did not find an open PR implementing this CSA multistream decode scheduling path.

## Correctness

Local import proof:
`vllm.__file__=/shared/amdgpu/home/fai_qle/vllm/vllm/__init__.py`

Full GSM8K 1319-question local-chat-completions run:
- strict-match: `0.9613`
- flexible-extract: `0.9606`

Additional checks:
- `.venv/bin/python -m pytest tests/models/test_deepseek_v4_rocm_multistream.py -q`: 7 passed
- `.venv/bin/python -m pytest tests/kernels/test_fused_deepseek_v4_qnorm_rope_kv_insert.py::test_split_q_and_kv_match_combined -q`: 12 passed
- `.venv/bin/python -m pytest tests/kernels/test_fused_deepseek_v4_qnorm_rope_kv_insert.py::test_kv_path_matches_reference -q -k 'not 2048'`: 8 passed, 2 deselected
- `.venv/bin/python -m py_compile vllm/models/deepseek_v4/nvidia/ops/attention.py vllm/v1/attention/ops/rocm_aiter_mla_sparse.py`: passed
- `git diff --check`: passed

## Benchmark: This PR vs InferenceX Baseline

Baseline: official InferenceX run, TP=8, fp8 KV, async scheduling, no prefix cache, `FULL_AND_PIECEWISE`, AITER enabled, `random_range_ratio=0.8`.

Lower TPOT/TTFT is better.

| case | c | out/gpu base | out/gpu this PR | out delta | TPOT base/PR ms | TPOT delta | TTFT base/PR s | TTFT delta |
|---|---:|---:|---:|---:|---:|---:|---:|---:|
| 1k/1k | 4 | 9.57 | 9.70 | +1.39% | 49.94/49.53 | -0.81% | 0.583/0.326 | -44.19% |
| 1k/1k | 8 | 18.31 | 18.68 | +2.05% | 52.65/51.91 | -1.39% | 0.672/0.365 | -45.63% |
| 1k/1k | 16 | 34.35 | 35.05 | +2.05% | 56.04/55.24 | -1.44% | 0.745/0.420 | -43.67% |
| 1k/1k | 32 | 60.06 | 61.53 | +2.46% | 64.00/62.53 | -2.30% | 0.628/0.564 | -10.20% |
| 1k/1k | 64 | 98.42 | 100.33 | +1.93% | 77.75/76.34 | -1.82% | 0.820/0.727 | -11.39% |
| 1k/1k | 128 | 69.33 | 70.50 | +1.69% | 225.80/222.10 | -1.64% | 1.398/1.280 | -8.44% |
| 1k/1k | 256 | 201.49 | 210.33 | +4.39% | 151.84/145.27 | -4.33% | 1.935/1.843 | -4.76% |
| 1k/1k | 512 | 273.17 | 283.76 | +3.88% | 224.80/216.25 | -3.80% | 3.447/3.343 | -3.02% |
| 8k/1k | 4 | 8.41 | 8.55 | +1.63% | 56.48/55.72 | -1.35% | 1.377/1.222 | -11.22% |
| 8k/1k | 8 | 15.47 | 15.68 | +1.39% | 61.70/60.96 | -1.19% | 1.650/1.528 | -7.38% |
| 8k/1k | 16 | 26.22 | 26.67 | +1.74% | 71.50/70.31 | -1.65% | 2.022/1.942 | -3.93% |
| 8k/1k | 32 | 41.35 | 41.97 | +1.52% | 90.99/89.58 | -1.55% | 2.812/2.783 | -1.05% |
| 8k/1k | 64 | 58.02 | 58.88 | +1.49% | 130.31/128.32 | -1.53% | 4.577/4.570 | -0.15% |
| 8k/1k | 128 | 48.11 | 48.75 | +1.33% | 320.15/315.88 | -1.33% | 8.126/8.056 | -0.86% |
| 8k/1k | 256 | 77.23 | 82.73 | +7.11% | 391.23/365.17 | -6.66% | 15.977/14.993 | -6.16% |
| 8k/1k | 512 | 89.46 | 91.16 | +1.91% | 675.30/662.19 | -1.94% | 30.901/30.942 | +0.13% |

## Notes

AI assistance was used to help implement, test, benchmark, and draft this PR.

### CODE REVIEW

**Comment by Fangzhou-Ai** [comment]
Hi @dllehr-amd can you please take a look at this PR. We enabled Multi-stream CSA here for better TTFT and TPOT. CC @ChuanLi1101 @wuhuikx

**Comment by ChuanLi1101** [comment]
Thanks for the thorough work on ROCm DSV4 CSA multistream decode — the split q/kv kernels, active-gating fix, and stream/event ordering improvements all look like the right direction.

A few small things before merge:

1. **Default opt-in**: Given the earlier hang/regression history, consider defaulting `VLLM_ROCM_DSV4_CSA_MULTISTREAM` to `off` until broader SKU/CI coverage, or call out opt-in clearly in docs.

2. **PR description vs commits**: The body table (1–7% gains) doesn't quite match the corrected numbers in `da077df` (+2% output / clearer TTFT win). Worth aligning the description with the final benchmark story and baseline (InferenceX official).

3. **`rocm_aiter_mla_sparse`**: Mentioned in the summary but not in the file list — either add the change or drop it from the description.

4. **Squash commits**: 11 commits with a long experimental arc — squashing to a few logical commits would help reviewers a lot.

5. **`multi_stream_utils`**: Changes affect non-ROCm callers too (e.g. LoRA) — a short note on CUDA smoke / expected impact would be helpful.

6. **Repro docs**: If `rocm_dsv4_stream_probe` was removed, keeping `tools/rocm_multistream_graph_repro.py` (or a brief design note) would help others reproduce the graph-capture findings.

Overall looks good to me once defaults and the PR narrative are tightened up. Happy to take another look after a rebase/squash.

File: vllm/models/deepseek_v4/nvidia/model.py [hunk 1]
```diff
@@ -8,6 +8,7 @@
 import torch
 import torch.nn as nn
 
+import vllm.envs as envs
```
Comment (tjtanaa) [inline vllm/models/deepseek_v4/nvidia/model.py]: We should not modify nvidia's modeling file. Please rebase your code and follow the state of the new code.

Modify only `vllm/models/deepseek_v4/amd/model.py`

File: vllm/envs.py [hunk 2]
```diff
@@ -256,6 +256,20 @@
     VLLM_DISABLE_SHARED_EXPERTS_STREAM: bool = False
     VLLM_SHARED_EXPERTS_STREAM_TOKEN_THRESHOLD: int = 256
     VLLM_MULTI_STREAM_GEMM_TOKEN_THRESHOLD: int = 1024
+    VLLM_ROCM_DSV4_CSA_MULTISTREAM: bool = True
```
Comment (tjtanaa) [inline vllm/envs.py]: Do not add of these flags. Please upstream the best configurations into this PR. Write it into code.

**Review by tjtanaa — CHANGES_REQUESTED** [review]

File: tests/models/test_deepseek_v4_rocm_multistream.py [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+
+
+def test_deepseek_v4_rocm_aux_streams_enabled(monkeypatch):
+    streams = [object(), object(), object(), object(), object()]
+
+    def make_stream(**kwargs):
+        assert kwargs == {"priority": -1}
+        return streams.pop()
+
+    monkeypatch.setattr(dsv4_model.current_platform, "is_rocm", lambda: True)
```
Comment (tjtanaa) [inline tests/models/test_deepseek_v4_rocm_multistream.py]: Please help to revise all of the unit tests generated from AI. 

For example, we don't need monkeypatch the  `current_platform`

Just use `current_platform.is_rocm` after `from vllm.platforms import current_platform` as usual. The unit test only runs on environment that has ROCm.

File: vllm/models/deepseek_v4/compressor.py [hunk 4]
```diff
@@ -242,13 +244,18 @@ def __init__(
             assert not use_fp4_cache, (
                 "MXFP4 cache is only supported for indexer (head=128)"
             )
-            self._use_cutedsl_sparse_compressor = True
-            self._use_cutedsl_fused_sparse_compressor = self.compress_ratio == 4
-            self._compress_kernel = _compress_kv_sparse_attn_cutedsl
-            self._norm_rope_store_kernel = _norm_rope_insert_sparse_attn_cutedsl
-            self._fused_sparse_kernel = (
-                _fused_kv_compress_norm_rope_insert_sparse_attn_cutedsl
+            self._use_cutedsl_sparse_compressor = (
```
Comment (tjtanaa) [inline vllm/models/deepseek_v4/compressor.py]: @Fangzhou-Ai this change is not needed anymore after you rebase with main, this import error is fixed in this PR https://github.com/vllm-project/vllm/pull/43710

File: tests/kernels/test_fused_deepseek_v4_qnorm_rope_kv_insert.py [hunk 5]
```diff
@@ -136,6 +136,22 @@ def _call_fused(
     )
 
 
+def _split_ops_available() -> bool:
+    return hasattr(torch.ops._C, "deepseek_v4_qnorm_rope") and hasattr(
```
Comment (tjtanaa) [inline tests/kernels/test_fused_deepseek_v4_qnorm_rope_kv_insert.py]: There is no point creating a wrapper function. And this test files only runs on platforms that compile this op. So we don't need to check this.

File: tests/kernels/test_fused_deepseek_v4_qnorm_rope_kv_insert.py [hunk 6]
```diff
@@ -136,6 +136,22 @@ def _call_fused(
...
+        torch.ops._C, "deepseek_v4_kv_rope_quant_insert"
+    )
+
+
+def _call_q_split(q, positions, cos_sin_cache, eps):
```
Comment (tjtanaa) [inline tests/kernels/test_fused_deepseek_v4_qnorm_rope_kv_insert.py]: There is no point creating a wrapper function.

File: tests/kernels/test_fused_deepseek_v4_qnorm_rope_kv_insert.py [hunk 7]
```diff
@@ -136,6 +136,22 @@ def _call_fused(
...
+    torch.ops._C.deepseek_v4_qnorm_rope(q, positions, cos_sin_cache, eps)
+
+
+def _call_kv_split(kv, k_cache, slot_mapping, positions, cos_sin_cache, bs):
```
Comment (tjtanaa) [inline tests/kernels/test_fused_deepseek_v4_qnorm_rope_kv_insert.py]: There is no point creating a wrapper function.

File: tests/kernels/test_fused_deepseek_v4_qnorm_rope_kv_insert.py [hunk 8]
```diff
@@ -415,3 +431,58 @@ def test_combined_q_and_kv(
             "padded head slots must be exact zero"
         )
     torch.testing.assert_close(k_cache_fused, k_cache_ref, rtol=0, atol=0)
+
+
+@pytest.mark.skipif(
+    not _split_ops_available(),
```
Comment (tjtanaa) [inline tests/kernels/test_fused_deepseek_v4_qnorm_rope_kv_insert.py]: If the op is not compiled on CUDA, and only runs on ROCm, use this skip condition instead, `not current_platform.is_rocm()`.

File: vllm/models/deepseek_v4/amd/model.py [hunk 9]
```diff
@@ -616,16 +642,7 @@ def __init__(self, *, vllm_config: VllmConfig, prefix: str = ""):
...
-        # DeepseekV4MultiHeadLatentAttentionWrapper.attn_gemm_parallel_execute
-        # (compressor kv_score, indexer.weights_proj, indexer.compressor
-        # kv_score). fused_wqa_wkv stays on the default stream.
-        # Disable them on ROCm because of hang issues.
-        aux_stream_list = (
-            None
-            if current_platform.is_rocm()
-            else [torch.cuda.Stream() for _ in range(3)]
-        )
+        aux_stream_list = make_deepseek_v4_aux_streams()
```
Comment (tjtanaa) [inline vllm/models/deepseek_v4/amd/model.py]: Don't need to wrap into the abstraction. Moreover, this modeling file is AMD exclusive, directly bake in the AMD logic here.

```
    return [
        torch.cuda.Stream(priority=_ROCM_DSV4_CSA_AUX_STREAM_PRIORITY)
        for _ in range(_ROCM_DSV4_CSA_AUX_STREAM_COUNT)
    ]
```

File: vllm/models/deepseek_v4/attention.py [hunk 10]
```diff
@@ -75,6 +79,47 @@
 
 logger = init_logger(__name__)
 
+_ROCM_DSV4_CSA_MS_STRATEGY = "overlap"
```
Comment (tjtanaa) [inline vllm/models/deepseek_v4/attention.py]: @Fangzhou-Ai Clean up all of this conditions, bake into the logic itself and we always enable the multistream.

File: vllm/models/deepseek_v4/nvidia/model.py (hunk 1 above)
Comment (Fangzhou-Ai) [reply vllm/models/deepseek_v4/nvidia/model.py]: Thanks for your comment! This line has been removed.

File: vllm/envs.py (hunk 2 above)
Comment (Fangzhou-Ai) [reply vllm/envs.py]: Thanks for your comment! All ENV knobs for tuning the multi-stream topology have been removed.

File: tests/models/test_deepseek_v4_rocm_multistream.py (hunk 3 above)
Comment (Fangzhou-Ai) [reply tests/models/test_deepseek_v4_rocm_multistream.py]: The test file has been re-written into the ROCm-only style and only those are directly linked to the multistream decoding path are kept.

File: vllm/models/deepseek_v4/compressor.py (hunk 4 above)
Comment (Fangzhou-Ai) [reply vllm/models/deepseek_v4/compressor.py]: This PR has been rebased and this file no longer needs modification.

File: tests/kernels/test_fused_deepseek_v4_qnorm_rope_kv_insert.py (hunk 5 above)
Comment (Fangzhou-Ai) [reply tests/kernels/test_fused_deepseek_v4_qnorm_rope_kv_insert.py]: Wrapper removed here.

File: tests/kernels/test_fused_deepseek_v4_qnorm_rope_kv_insert.py (hunk 6 above)
Comment (Fangzhou-Ai) [reply tests/kernels/test_fused_deepseek_v4_qnorm_rope_kv_insert.py]: Wrapper removed here

File: tests/kernels/test_fused_deepseek_v4_qnorm_rope_kv_insert.py (hunk 7 above)
Comment (Fangzhou-Ai) [reply tests/kernels/test_fused_deepseek_v4_qnorm_rope_kv_insert.py]: Wrapper removed here

File: tests/kernels/test_fused_deepseek_v4_qnorm_rope_kv_insert.py (hunk 8 above)
Comment (Fangzhou-Ai) [reply tests/kernels/test_fused_deepseek_v4_qnorm_rope_kv_insert.py]: The helper function is removed.

File: vllm/models/deepseek_v4/amd/model.py (hunk 9 above)
Comment (Fangzhou-Ai) [reply vllm/models/deepseek_v4/amd/model.py]: The helper function has been directly re-written into the inline function style on ROCm path.

File: vllm/models/deepseek_v4/attention.py (hunk 10 above)
Comment (Fangzhou-Ai) [reply vllm/models/deepseek_v4/attention.py]: All extra ENV knobs are removed with fixed optimal config baked into the code.

File: vllm/utils/multi_stream_utils.py
Comment (zyongye) [inline vllm/utils/multi_stream_utils.py:1]: Why do we need to change this file?

**Review by zyongye — COMMENTED** [review]
If we introduce too much CUDA/ROCM divergence. We should consider split this file and only perform it in ROCM specific branch instead.

**Comment by Fangzhou-Ai** [comment]
> If we introduce too much CUDA/ROCM divergence. We should consider split this file and only perform it in ROCM specific branch instead. 

Thanks for your comment. This is a good idea, I'll separate the changes into a more explicit way.

**Comment by zyongye** [comment]
> > If we introduce too much CUDA/ROCM divergence. We should consider split this file and only perform it in ROCM specific branch instead.
> 
> Thanks for your comment. This is a good idea, I'll separate the changes into a more explicit way.

Thank you for the response. Still I wonder is there any difference between doing multi-stream between these two platform?
